pkg/exchange/ftx/stream.go

```diff
@@ -102,15 +117,69 @@ func (s *Stream) SetPublicOnly() {
 	atomic.StoreInt32(&s.publicOnly, 1)
 }
 
-func (s *Stream) Subscribe(channel types.Channel, symbol string, _ types.SubscribeOptions) {
-	if channel != types.BookChannel {
-		panic("only support book channel now")
+func (s *Stream) Subscribe(channel types.Channel, symbol string, option types.SubscribeOptions) {
+	if channel == types.BookChannel {
+		s.addSubscription(websocketRequest{
+			Operation: subscribe,
+			Channel:   orderBookChannel,
+			Market:    toLocalSymbol(TrimUpperString(symbol)),
+		})
+
+	} else if channel == types.KLineChannel {
+		// FTX does not support kline channel, do polling
+		go s.subscribeKLine(symbol, option)
+	} else {
+		panic("only support book/kline channel now")
 	}
-	s.addSubscription(websocketRequest{
-		Operation: subscribe,
-		Channel:   orderBookChannel,
-		Market:    toLocalSymbol(TrimUpperString(symbol)),
+}
+
+func (s *Stream) handleChannelKlineMessage(ctx context.Context) {
+	for {
+		select {
+		case km := <-s.km:
+			// FTX only returns closed kline
+			klines := getLastKLine(s.exchange, ctx, km.symbol, km.interval)
+			if len(klines) >= 0 {
+				// handle mutiple klines, get the latest one
+				s.EmitKLineClosed(klines[len(klines)-1])
+			}
+		}
+	}
+}
+
+func (s *Stream) subscribeKLine(symbol string, option types.SubscribeOptions) {
+	interval := types.Interval(option.Interval)
+	if !isIntervalSupportedInKLine(interval) {
+		logger.Errorf("not supported kline interval %s", option.Interval)
+		return
+	}
+
+	// get current kline candle
+	s.km <- klineMessage{symbol: symbol, interval: interval}
+
+	ticker := time.NewTicker(interval.Duration())
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			s.km <- klineMessage{symbol: symbol, interval: interval}
+		}
+	}
+}
+
+func getLastKLine(e *Exchange, ctx context.Context, symbol string, interval types.Interval) []types.KLine {
+	// set since to more 30s ago to avoid getting no kline candle
+	since := time.Now().Add(time.Duration(-1*(interval.Minutes()*60+30)) * time.Second)
```

Comment:
how about using time.Truncate and add -time.Millisecond

Reply:
Because the current timestamp may be `interval+N ms`, and we need klines from the previous time slot, not the current one. The formula should be:

`since := time.Now().Truncate(interval.Duration()).Add(time.Duration(-1*(interval.Minutes()*60*1000)-1) * time.Millisecond)`

I guess it is not simpler than original one